components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java

```diff
@@ -0,0 +1,223 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ * under the License.
+ *
+ *
+ */
+
+package org.wso2.carbon.identity.user.account.association.handler.grant;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.common.model.User;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.OAuth2TokenValidationService;
+import org.wso2.carbon.identity.oauth2.ResponseHeader;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2ClientApplicationDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationRequestDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationResponseDTO;
+import org.wso2.carbon.identity.oauth2.model.RequestParameter;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+import org.wso2.carbon.identity.oauth2.token.handlers.grant.AbstractAuthorizationGrantHandler;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.identity.user.account.association.UserAccountConnectorImpl;
+import org.wso2.carbon.identity.user.account.association.dto.UserAccountAssociationDTO;
+import org.wso2.carbon.identity.user.account.association.exception.UserAccountAssociationException;
+
+import java.util.Arrays;
+import java.util.HashSet;
+
+public class AccountSwitchGrantHandler extends AbstractAuthorizationGrantHandler {
+
+    private static final Log log = LogFactory.getLog(AccountSwitchGrantHandler.class);
+
+    public static final String TOKEN_PARAM = "token";
+    public static final String USERNAME_PARAM = "username";
+    public static final String USERSTORE_DOMAIN_PARAM = "userstore-domain";
+    public static final String TENANT_DOMAIN_PARAM = "tenant-domain";
+    private final String OAUTH_HEADER = "Bearer";
+    private final String ACCOUNT_SWITCH_TOKEN_VALIDATION_CONTEXT = "accountSwitchTokenValidationContext";
+
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        super.validateGrant(tokReqMsgCtx);
+
+        String token = extractParameter(TOKEN_PARAM, tokReqMsgCtx);;
+        String username = extractParameter(USERNAME_PARAM, tokReqMsgCtx);;
+        String userstoreDomain = extractParameter(USERSTORE_DOMAIN_PARAM, tokReqMsgCtx);;
+        String tenantDomain = extractParameter(TENANT_DOMAIN_PARAM, tokReqMsgCtx);;
+
+        if (StringUtils.isEmpty(token) || StringUtils.isEmpty(username) || StringUtils.isEmpty(username) ||
+                StringUtils.isEmpty(username)) {
```

Comment:
Change last two conditions to || StringUtils.isEmpty(**userstoreDomain**) || StringUtils.isEmpty(**tenantDomain**) {

Reply:
Oops

Reply:
And we don't need these we have AccountSwitchGrantValidator.

Reply:
Validated the flow.

GrantValidator doing the validation of the parameters and sending back `400 Bad Request` response with something like following content,
```
{
    "error_description": "OAuthProblemException{error='invalid_request', description='Missing parameters: tenant-domain token', uri='null', state='null', scope='null', redirectUri='null', responseStatus=0, parameters={}}",
    "error": "invalid_request"
}
```